Fix Issue 1716: Split customized edgex-vault docker into two docker images

jim-wang-intel opened the pull request:
This PR fix the issue #1716 

The original customized edgex-vault docker image is broken into two separate images:
 - the official Vault docker image from HashiCorp
 - the security-secrets-setup docker image

The testing example of docker-compose yml file now contains two separate [vault official docker image](https://github.com/jim-wang-intel/edgex-go/blob/issue-1716/cmd/security-secrets-setup/testdata/test-example-security.yml#L84-L134) and [security-secrets-setup edgex docker image](https://github.com/jim-wang-intel/edgex-go/blob/issue-1716/cmd/security-secrets-setup/testdata/test-example-security.yml#L136-L150).

- docker vault service command line has script to wait for the PKI initialization to be complete,
using multiple line YAML indicator for easier readability

- docker edgex-secrets-setup image has new environment variable `VAULT_LOCAL_CONFIG` for hooking up the TLS assets for vault and is using multiple line YAML indicator for easier readability

- docker volume mapping `secrets-setup-cache:/etc/edgex/pki` is added into compose file for PKI cache use

To run different subcommands of `security-secrets-setup`, one can edit the docker-compose file near
the `command` section of `security-secrets-setup` docker service.
One can comment and un-comment out [the provided templates](https://github.com/jim-wang-intel/edgex-go/blob/issue-1716/cmd/security-secrets-setup/edgex-stack-security.yml#L141-L144) for `generate`, `cache`, and `import`
in the compose-file to run the desired mode of operation respectively.

Ref:
 - Issue #1716 

Signed-off-by: Jim Wang <yutsung.jim.wang@intel.com>

tingyuz commented on cmd/security-secrets-setup/edgex-stack-security.yml:
```diff
@@ -0,0 +1,583 @@
+#  * the License.
+#  *
+#  * @author: Jim White, Dell
+#  * EdgeX Foundry, Edinburgh, version 1.0.1
```
Per the discussion during core working group in the past, it is suggested to remove the version number from the header file. If a version number is needed, it is better to put into the file name.

jim-wang-intel replied:
yes, this is just a testing example. will remove it

tingyuz commented on cmd/security-secrets-setup/edgex-stack-security.yml:
```diff
@@ -0,0 +1,583 @@
+      - consul
+
+  security-secrets-setup:
+    image: edgexfoundry/docker-edgex-vault:1.1.0-dev
```
-dev suffix will not be in the merged version. Suggest to user : 1.1.0 instead.

jim-wang-intel replied:
this is just a testing example- in the proposed fuji docker-compose will use the version without -dev.  BTW, what is the process for publishing the docker image to the edgexfoundry docker hub or repo? @tingyuz

tingyuz commented on cmd/security-secrets-setup/edgex-stack-security.yml:
```diff
@@ -0,0 +1,583 @@
+      - volume
+
+  vault-worker:
+    image: edgexfoundry/docker-edgex-vault-worker-go:1.0.0
```
may be 1.1.0 here?

tingyuz commented on cmd/security-secrets-setup/edgex-stack-security.yml:
```diff
@@ -0,0 +1,583 @@
+    image: edgexfoundry/docker-edgex-vault-worker-go:1.0.0
+    container_name: edgex-vault-worker
+    hostname: edgex-vault-worker
+    command: ["--init=true", "--debug=false", "--wait=10", "--insureskipverify=false"]
```
This line needs to be command: ["--init=true", "--debug=false", "--vault-interval=10", "--insecurekipverify=false"] .
We updated the vault-worker but not this compose file, so hope this can get in with the PR as well.

mkbhanda replied:
--insecureskipverify=false

tingyuz commented on cmd/security-secrets-setup/edgex-stack-security.yml:
```diff
@@ -0,0 +1,583 @@
+        - kong-db
+
+  edgex-proxy:
+    image: "edgexfoundry/docker-edgex-proxy-go:1.0.0"
```
this will be 1.1.0 as well.

bnevis-i commented:
I'm thinking rather than getting mired in debating the innacuracies of `edgex-stack-security.yml` , we should remove it from this PR and modify the official docker compose to really integrate this work -- is that [https://github.com/edgexfoundry/developer-scripts/blob/master/releases/nightly-build/compose-files/docker-compose-nexus.yml](https://github.com/edgexfoundry/developer-scripts/blob/master/releases/nightly-build/compose-files/docker-compose-nexus.yml) ???

lenny-intel commented:
Not that I put this PR on hold per this weeks WG decision to push this to the next release.

tingyuz commented:
@bnevis-i @jim-wang-intel  The official docker-compose file for each release will be residing here:
https://github.com/edgexfoundry/developer-scripts/tree/master/releases

In the past release we created a docker-compose file under our own project for now, then we updated the docker-compose file under developer-scripts folder based on the file in each individual project as each project may only update the related entries in the docker-compose file.  If we have this docker-compose merged here I will follow up to update the master docker-compose file in developer-scripts folder.

bnevis-i commented:
> @bnevis-i @jim-wang-intel The official docker-compose file for each release will be residing here:
> https://github.com/edgexfoundry/developer-scripts/tree/master/releases
> 
> In the past releases we created a docker-compose file under our own project, then we updated the docker-compose file under developer-scripts folder based on the file in each individual project as each project may only update the related entries in the docker-compose file. If we have this docker-compose merged here I will follow up to update the master docker-compose file in developer-scripts folder.

So maybe the right thing to do is fork developer-scripts, make a pull request that proposes a fuji release docker-compose, and then refer that that script in the testing instructions?

tingyuz commented:
> > @bnevis-i @jim-wang-intel The official docker-compose file for each release will be residing here:
> > https://github.com/edgexfoundry/developer-scripts/tree/master/releases
> > In the past releases we created a docker-compose file under our own project, then we updated the docker-compose file under developer-scripts folder based on the file in each individual project as each project may only update the related entries in the docker-compose file. If we have this docker-compose merged here I will follow up to update the master docker-compose file in developer-scripts folder.
> 
> So maybe the right thing to do is fork developer-scripts, make a pull request that proposes a fuji release docker-compose, and then refer that that script in the testing instruction?


I think we can do both - just working on the docker-compose file here for our testing, and open another PR against developer-scripts folder to work with other teams to update the master dockercompose file. 

@jim-wang-intel @badboy-huaqiao please try if you guys have write permission against developer-script project/folder, or we need to request it through LF DevOps.

anonymouse64 commented:
Sorry missed a number of comments before submitting my review...

@tingyuz 

> I think we can do both

I strongly suggest we don't do both since there's always a problem of updating two things and keeping them in sync. Since we have already historically put compose files in developer-scripts, I'd suggest they stay there. Also note I already filed an issue for this against the old security-secret-store repo: https://github.com/edgexfoundry/security-secret-store/issues/63

@bnevis-i 

> So maybe the right thing to do is fork developer-scripts, make a pull request that proposes a fuji release docker-compose, and then refer that that script in the testing instructions?

I'd say yes

tingyuz commented:
> Didn't we agree in the WG not to do this until after Fuji and other higher priority things were done?
> 
> I'm suggesting we either tag this for Geneva / Fuji dot or close and re-open this when we are after the Fuji release date.
> 
> Also, fwiw historically I think docker-compose files lived in developer-scripts instead of edgex-go.

My understanding is that since new pki-init has been merged we will need new docker-compose file to create cert on the fly, or the old docker-compose file will not work. Correct me if I am wrong here. @jim-wang-intel @bnevis-i

anonymouse64 commented:
> we will need new docker-compose file to create cert on the fly, or the old docker-compose file will not work

Sure, but why does that necessitate putting the docker-compose file in this repo? Wouldn't we just add a docker-compose-fuji-security.yaml and a docker-compose-fuji-no-security.yaml as part of the Fuji work/release process?
Perhaps I am misunderstanding, are you proposing that we need a third docker-compose-fuji-security-other-thing.yaml?

jim-wang-intel commented:
> 
> 
> I'm thinking rather than getting mired in debating the innaccuracies of `edgex-stack-security.yml` , we should remove it from this PR and modify the official docker compose to really integrate this work -- the nightly [nightly](https://github.com/edgexfoundry/developer-scripts/blob/master/releases/nightly-build/compose-files/docker-compose-nexus.yml) doesn't seem to be the right one. @tingyuz do you know where the official working docker-compose is kept -- for Fuji?

The existing nightly one does not seem work, having trouble to pull the mongo docker image...I think there is some issue for that nightly docker-compose file...

jim-wang-intel commented:
@anonymouse64 , @lenny-intel , @tingyuz  this PR is just trying to resolve the issue #1716; regarding to the release version of docker-compose files whether it is for Fuji or Fuji-dot or even beyond, I'll leave it up to the community's final decision and they will be in developer-script repo.  The docker-compose file here is just for testing example purpose; so I renamed it and moved it under `./testdata`.

jim-wang-intel commented:
Per conversation of Lenny and Ian, this PR is on hold and closed
